### PR TITLE
fix(platform/pr-body): refactor smartTruncate function

### LINKED ISSUE #14551: Pull request creation error - body is too long (maximum is 65536 characters)
### How are you running Renovate?

WhiteSource Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

_No response_

### Please select which platform you are using if self-hosting.

_No response_

### If you're self-hosting Renovate, tell us what version of the platform you run.

_No response_

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

Renovate fails to create PR when using `groupName`. See the attached logs for more details

Screenshot 2022-03-07 at 11 16 39

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Pull request creation error(branch="renovate/hello-world")
{
  "err": {
    "hostType": "github",
    "err": {
      "name": "HTTPError",
      "code": "ERR_NON_2XX_3XX_RESPONSE",
      "timings": {
        "start": 1646643660083,
        "socket": 1646643660084,
        "lookup": 1646643660084,
        "connect": 1646643660095,
        "secureConnect": 1646643660106,
        "upload": 1646643660107,
        "response": 1646643660585,
        "end": 1646643660585,
        "phases": {
          "wait": 1,
          "dns": 0,
          "tcp": 11,
          "tls": 11,
          "request": 1,
          "firstByte": 478,
          "download": 0,
          "total": 502
        }
      },
      "message": "Response code 422 (Unprocessable Entity)",
      "stack": "HTTPError: Response code 422 (Unprocessable Entity)\n    at Request.<anonymous> (/home/ubuntu/renovateapp/node_modules/got/dist/source/as-promise/index.js:117:42)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)",
      "options": {
        "headers": {
          "user-agent": "Renovate Bot (GitHub App 2740)",
          "accept": "application/vnd.github.machine-man-preview+json",
          "authorization": "***********",
          "content-type": "application/json",
          "content-length": "73385",
          "accept-encoding": "gzip, deflate, br"
        },
        "url": "https://api.github.com/repos/mentimeter/mm-js/pulls",
        "hostType": "github",
        "username": "",
        "password": "",
        "method": "POST",
        "http2": false
      },
      "response": {
        "statusCode": 422,
        "statusMessage": "Unprocessable Entity",
        "body": {
          "message": "Validation Failed",
          "errors": [
            {
              "resource": "Issue",
              "code": "custom",
              "field": "body",
              "message": "body is too long (maximum is 65536 characters)"
            }
          ],
          "documentation_url": "https://docs.github.com/rest/reference/pulls#create-a-pull-request"
        },
        "headers": {
          "server": "GitHub.com",
          "date": "Mon, 07 Mar 2022 09:01:00 GMT",
          "content-type": "application/json; charset=utf-8",
          "content-length": "242",
          "x-github-media-type": "github.v3; param=machine-man-preview; format=json",
          "x-ratelimit-limit": "8900",
          "x-ratelimit-remaining": "8861",
          "x-ratelimit-reset": "1646645396",
          "x-ratelimit-used": "39",
          "x-ratelimit-resource": "core",
          "access-control-expose-headers": "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset",
          "access-control-allow-origin": "*",
          "strict-transport-security": "max-age=31536000; includeSubdomains; preload",
          "x-frame-options": "deny",
          "x-content-type-options": "nosniff",
          "x-xss-protection": "0",
          "referrer-policy": "origin-when-cross-origin, strict-origin-when-cross-origin",
          "content-security-policy": "default-src 'none'",
          "vary": "Accept-Encoding, Accept, X-Requested-With",
          "x-github-request-id": "8CF2:3E4A:8E14B2:93A218:6225C9CC",
          "connection": "close"
        },
        "httpVersion": "1.1"
      }
    },
    "message": "external-host-error",
    "stack": "Error: external-host-error\n    at handleGotError (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/http/github.js:92:16)\n    at GithubHttp.request (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/http/github.js:219:19)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async GithubHttp.requestJson (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/http/index.js:149:21)\n    at async Proxy.createPr (/home/ubuntu/renovateapp/node_modules/renovate/dist/platform/github/index.js:1174:17)\n    at async ensurePr (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/pr/index.js:349:22)\n    at async processBranch (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/branch/index.js:497:32)\n    at async writeUpdates (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/write.js:25:21)\n    at async update (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/extract-update.js:108:15)\n    at async Object.renovateRepository (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/index.js:41:21)\n    at async renovateRepository (/home/ubuntu/renovateapp/app/worker/index.js:427:26)\n    at async /home/ubuntu/renovateapp/app/worker/index.js:678:5"
  }
}
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

### PR DESCRIPTION
**Pull request by MaronHatoum**
## Changes
in the `smartTruncate` function, escape characters are not counted and after the `replace` the escape characters were added to the returned value, as a result, we are getting a `prBody` length greater than the requested length. 
so I refactored the function which counts the requested length including the escape characters, therefore it returns exactly the 'prBody' length as requested.

## Context
Closes https://github.com/renovatebot/renovate/issues/14551

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/platform/utils/pr-body.ts [hunk 1]
```diff
@@ -1,23 +1,27 @@
-const re = new RegExp(`### Release Notes.*### Configuration`, 'ms');
+const re = new RegExp(`(.*### Release Notes)(.*)### Configuration(.*)`, 'ms');
```
Comment (viceice) [inline lib/modules/platform/utils/pr-body.ts:1]: Use named capture groups and use `regEx` from `util/regex`

Comment (MaronHatoum) [reply lib/modules/platform/utils/pr-body.ts:1]: Hi @viceice 
I used capture groups 
`const re = regEx(`(?<preNotes>.*### Release Notes)(?<releaseNotes>.*)### Configuration(?<postNotes>.*)`,'ms');`
but when executing `re.exec(input)` it returns null 
and when using the same pattern with new RegExp
`const re =new RegExp(`(?<preNotes>.*### Release Notes)(?<releaseNotes>.*)### Configuration(?<postNotes>.*)`,'ms');` 
it works fine.
should I change the pattern?

Comment (viceice) [reply lib/modules/platform/utils/pr-body.ts:1]: I think the `m` flag is wrong. it causes `^` and `$` to match start and end of line, so can you test without `m` flag and only use `s` flag?

Comment (MaronHatoum) [reply lib/modules/platform/utils/pr-body.ts:1]: @viceice I removed `m` flag and it doesn't work (`re.exec(input)` returns null).

**Review by rarkins — DISMISSED** [review]

**Review by viceice — APPROVED** [review]
LGTM, please mark as ready before (re-)request review :)

**Pull request merged**